### ISSUE TITLE
[Postgres] Handling NULL bytes

### ISSUE THREAD

**Issue by frogcjn** [sequelize/sequelize#6485]
SQL:

``` sql
INSERT INTO "content"(
    "name"
)
VALUES
    (
        'happy\u0000',
    );
```

**Comment by sushantdhiman** [comment]
Please provide a SSCCE ( [What is SSCCE ?](http://sscce.org), [How to create one in Sequelize ?](https://github.com/sequelize/sequelize/blob/master/sscce_template.js) )

**Comment by frogcjn** [comment]
``node
var Maya = this.sequelize.define('Maya', {
        info: Sequelize.TEXT
      });
await Maya.create({info: JSON.strigify({name: "1234\u0000"}))
```

**Comment by felixfbecker** [comment]
Seems like Sequelize doesn't double-escape the NUL byte.

**Comment by frogcjn** [comment]
await Maya.create({info: `1234\u0000`})

**Comment by felixfbecker** [comment]
Which version are you using? This problem will probably not occur anymore once we use prepared statements.

**Comment by frogcjn** [comment]
This problem only appears with postgres, not with mysql.
v4.0.0-1

**Comment by frogcjn** [comment]
Maybe this is an issue with pg.

**Comment by felixfbecker** [comment]
I don't think so, we still do the escaping ourselves in latest master. Btw, what is the SQL output you would expect?

**Comment by frogcjn** [comment]
INSERT 0 1

**Comment by frogcjn** [comment]
Related issue: https://github.com/sequelize/sequelize/issues/3444

**Comment by felixfbecker** [comment]
If you expect the input to be stripped, you should do it yourself. If a user wants to insert a NUL byte, he should be able to do so.

**Comment by frogcjn** [comment]
NUL byte is not allowed in postgres text field. I'll watch out.

**Comment by frogcjn** [comment]
@felixfbecker  how to add a hook for each TEXT field to strip NUL byte?

**Comment by felixfbecker** [comment]
#6290 could allow this by overriding `stringify()`. Atm you could define `beforeCreate`, `beforeUpdate` etc hooks. But imo you should better solve this on your application layer.

**Comment by frogcjn** [comment]
I think for postgres, there should add validate for \u0000

**Comment by frogcjn** [comment]
for others who encounter the same problem:
```node
sequelize.TEXT.prototype._stringify = (value: string, option: any) => {
    return stripNullByte(value)
}
```

**Comment by felixfbecker** [comment]
@frogcjn that is private API and _will_ break in a future release of 4.0

**Comment by frogcjn** [comment]
@felixfbecker 
JSONB may have another issue
```
    const Content = sequelize.define("content", {
            info: Sequelize.JSONB
        })
    await Content.sync({ force: true })
    await Content.create({info: `happy\u0000`})
```

{ SequelizeDatabaseError: unsupported Unicode escape sequence

**Comment by contra** [comment]
AFAIK This is actually pretty serious and might be a security issue.

**Comment by frogcjn** [comment]
@contra, it is serious problem. since PG not give a wonderful solution, it just throws an error, I think sequelize should do something. At least to validate the text, or replace \u0000 to "" for TEXT and JSONB fields, as the code I wrote.

**Comment by contra** [comment]
@frogcjn Not just for inserts, for all queries.

**Comment by frogcjn** [comment]
@contra, if you replace original _stringify of the data type, then all queries sql will not contains \u0000.
```node
        const Maya = sequelize.define("maya", {name: Sequelize.TEXT})
        await Maya.sync({force: true})
        await Maya.create({name: `happy\u0000`})
        await Maya.find({where: {name: `happy\u0000`}, logging: console.log})
```
This is safe:
```
Executing (default): SELECT "id", "name", "createdAt", "updatedAt" FROM "mayas" AS "maya" WHERE "maya"."name" = 'happy' LIMIT 1;
```

**Comment by felixfbecker** [comment]
The escaping is not done in `_stringify` AFAIK but in the `QueryGenerator`
@contra This only applies to NUL bytes not being escaped, right?

**Comment by contra** [comment]
@felixfbecker Yeah, NUL bytes should be escaped or removed before being sent to postgres - users shouldn't be able to cause message protocol errors.

**Comment by mickhansen** [comment]
NULL bytes are not allowed by Postgres in strings, use BYTEA if you want to store NULL bytes, Sequelize should not attempt to circumvent this.

If a user preferes reencoding null bytes they are welcome to do so. Sequelize should not give the option of a silent failure, you might aswell choose MySQL in that case ;)

As long as it does not present as an injection issue i say we let it be.

**Comment by frogcjn** [comment]
I think sequelize should at least tell developer this issue when it happens, or validate the TEXT /JSONB data, or give a link to this discussion, since the error given by pg is not very meaningful hints
(Invalid format or Unsupported Unicode escape sentence) these errors do not tell any useful information .

**Comment by frogcjn** [comment]
Give a solution to this problem is saving others time. Since this problem always happen when \u0000 appears.

**Comment by mickhansen** [comment]
@frogcjn I agree. Sequelize could improve developer experience by wrapping it into a better error.
Possibly via validation, but i don't think we should silently strip the values. NULL bytes in your data should not be all the common i'd say, unless you're dealing with mailicious input.

**Comment by felixfbecker** [comment]
Imo we should do a `indexOf('\0')` check in `stringify()` or something like that. I think `stringify()` should only accept valid arguments, and in the case of Postgres strings it's a JS string that doesn't contain null bytes.

**Comment by contra** [comment]
Should be reopened!

**Comment by martywins** [comment]
@mickhansen 
Have you considered making this behaviour configurable?
Eg. some users may want to strip out any strange characters like this prior to sending anything to postgres.

**Comment by contra** [comment]
Not stale, keep it open.

**Comment by ulisesgalvan** [comment]
??

**Comment by ulisesgalvan** [comment]
Still open?

**Comment by aheuermann** [comment]
Is there a solution for this other than updating the private api?

**Comment by aheuermann** [comment]
I'm running into this issue when using `JSONB` which contains the `\u0000` null byte character. I'm able to escape it with the extra `\`

`{ "a":"\\u0000"}` ***will insert*** fine while `{ "a":"\u0000"}` will ***fail***

**Comment by papb** [comment]
Hi everyone, can someone please clarify what is the desired behavior here? Should `\u0000` and `\\u0000` simply be stripped away? Thanks

**Comment by germanoeich** [comment]
@papb `\u0000` has no representation and no width, I think the safest behaviour would be to strip it. If replaced by \\u0000, it could bypass validations for alphanumeric only fields and cause obscure problems down the line.

Also upping this, would be a really nice addition. We handle data from users in a service that supports unicode and we've hit this issue when importing data from mongo to postgres using sequelize.

**Comment by ephys** [comment]
To add my two cents: I think we should handle null bytes, but how we handle them should be behind an option.

There should be a global option that let's you choose between `error` or `remove` when encountering a null byte
